client/src/connection/itc/index.ts

```diff
@@ -0,0 +1,387 @@
+    this._shellProcess.stdin.write(
+      `
+do {
+  $chunkSize = 32768
+  $log = $runner.FlushLog($chunkSize)
+  Write-Host $log
+} while ($log.Length -gt 0)\n
+  `,
+      this.onWriteComplete,
+    );
+  };
+
+  /**
+   * Handles stderr output from the powershell child process.
+   * @param chunk a buffer of stderr output from the child process.
+   */
+  private onShellStdErr = (chunk: Buffer): void => {
+    const msg = chunk.toString();
+    console.warn("shellProcess stderr: " + msg);
+    this.clearPassword();
+    this._runReject(
+      new Error(
+        "There was an error executing the SAS Program.\nSee console log for more details.",
+      ),
+    );
+    // If we encountered an error in setup, we need to go through everything again
+    if (/Setup error/.test(msg)) {
+      this._shellProcess.kill();
+      this._workDirectory = undefined;
+    }
+  };
+
+  /**
+   * Handles stdout output from the powershell child process.
+   * @param data a buffer of stdout output from the child process.
+   */
+  private onShellStdOut = (data: Buffer): void => {
+    const output = data.toString().trimEnd();
+    const outputLines = output.split(/\n|\r\n/);
+
+    outputLines.forEach((line: string) => {
+      if (!line) {
+        return;
+      }
+      if (!this._workDirectory && line.startsWith("WORKDIR=")) {
+        const parts = line.split("WORKDIR=");
+        this._workDirectory = parts[1].trim();
+        this._runResolve();
+        return;
+      }
+      if (line.endsWith(endCode)) {
+        // run completed
+        this.fetchResults();
+      } else {
+        this._html5FileName =
+          line.match(/NOTE: .+ HTML5.* Body.+: (.+)\.htm/)?.[1] ??
+          this._html5FileName;
+        this._onLogFn?.([{ type: "normal", line }]);
+      }
+    });
+  };
+
+  /**
+   * Generic call for use on stdin write completion.
+   * @param err The error encountered on the write attempt. Undefined if no error occurred.
+   */
+  private onWriteComplete = (err: Error): void => {
+    if (err) {
+      this._runReject?.(err);
+    }
+  };
+
+  /**
+   * Not implemented.
+   */
+  public sessionId = (): string => {
+    throw new Error("Not Implemented");
+  };
+
+  /**
+   * Fetches the ODS output results for the latest html results file.
+   */
+  private fetchResults = async () => {
+    const globalStorageUri = getGlobalStorageUri();
+    try {
+      await workspace.fs.readDirectory(globalStorageUri);
+    } catch (e) {
+      await workspace.fs.createDirectory(globalStorageUri);
+    }
+
+    const outputFileUri = Uri.joinPath(
+      globalStorageUri,
+      `${this._html5FileName}.htm`,
+    );
+    this._shellProcess.stdin.write(
+      `$filePath = "${resolve(
+        this._workDirectory,
+        this._html5FileName + ".htm",
+      )}"
+$outputFile = "${outputFileUri.fsPath}"
+$runner.FetchResultsFile($filePath, $outputFile)\n`,
+      this.onWriteComplete,
+    );
+
+    const file = await new Promise((resolve) => {
```

Comment:
Just wondering why were checking to see if maxTime has passed here?  Are we waiting that long for a failed read before giving up? Just wondering if we'd run up against that for potentially large files that might take longer than 10 seconds to read?

Reply:
Yeah, that's a good point, @smorrisj . The reason we have maxTime here is because `FetchResultsFile` takes _some_ time to store the results. 

I'm wondering what a better approach would look like (since we don't _really_ know when FetchResultsFile completes).

Reply:
Hmm. Could we invent a way to know when it's done? One such thing might be to have the powershell script echo out a unique line that we look for to know that we fetched results, then react to it if we see that line echoed to the onShellStdOut?

Reply:
Resolved in https://github.com/sassoftware/vscode-sas-extension/pull/592/commits/1da05d1a445dde2144bc471b9cc842da3dccebc4